Core/MMKV.cpp

```diff
@@ -436,6 +436,15 @@ bool MMKV::set(bool value, MMKVKey_t key, uint32_t expireDuration) {
     if (isKeyEmpty(key)) {
         return false;
     }
+
+    if (getCompareBeforeSetEnabled()) {
+        bool hasValue;
+        bool oldValue = getBool(key, false, &hasValue);
```

Comment:
No, it doesn't work this way. Passing a `bool` in doesn't guarantee that the existing value is also a `bool`. The other `set()` methods below have similar problems.

Reply:
Have been some confused.

There is nothing about accurate TYPE for the data in file, totally depends on the type passed by users.

Just like my tests
```
mmkv1->set("value1", "key1");
printf("%d\n", mmkv1->getBool("key1", false)); // print 1(true)
```

How to handle this cases? 😂

Reply:
Maybe you should compare the data AS IS.

Reply:
My first pr is to  compare the data AS IS, 
```
// compare data before appending to file
                if (m_enableCompareBeforeSet) {
                    MMBuffer rawData = getDataForKey(key);
                    if (isDataHolder) {
                    } else {
                    }
```

but I remember your mentioning about the incorrectness of `isDataHolder`

For the first and current pr, which one is suitable？ or none of them? 

Maybe you can give more hints, so I keep this moving on

Reply:
You should solve the `isDataHolder` problem on the base of comparing the data AS IT. This PR is in a totally wrong direction, I frankly don't understand why you go this way. It doesn't solve the `isDataHolder` problem at all. It also brings the problem to all data whether they are holders or not.